### ISSUE TITLE
Gradient exploding

### ISSUE THREAD

**Issue by ljj430** [eth-sri/SafeCoder#2]
Hi,

I am very interested in your work and tried to replicate it. However, I met gradient exploding issue, i.e., the loss becomes nan, when I tried the example from README
```
xxx@xxx:~/xxx/SafeCoder$ source activate secure_oss && CUDA_VISIBLE_DEVICES=0 python scripts/train.py --pretrain_name starcoderbase-1b --output_name starcoderbase-1b-safecoder --datasets evol sec-desc sec-new-desc
The `load_in_4bit` and `load_in_8bit` arguments are deprecated and will be removed in the future versions. Please, pass a `BitsAndBytesConfig` object in `quantization_config` argument instead.
03/12/2025 14:27:03 - INFO - accelerate.utils.modeling -   We will use 90% of the memory on device 0 for storing the model, and 10% for the buffer to avoid OOM. You can set `max_memory` in to a higher value to use more memory (at your own risk).
03/12/2025 14:27:44 - INFO - root -   number of sec samples before upsampling: 1848
03/12/2025 14:27:44 - INFO - root -   number of sec samples after upsampling: 2468
03/12/2025 14:27:49 - INFO - root -   Training args Namespace(output_name='starcoderbase-1b-safecoder', datasets=['evol', 'sec-desc', 'sec-new-desc'], pretrain_name='starcoderbase-1b', loss_weight=1.0, sven=False, num_train_epochs=2, learning_rate=2e-05, max_num_tokens=1024, batch_size=1, grad_acc_steps=16, weight_decay=0.01, adam_epsilon=1e-08, warmup_steps=0, max_grad_norm=1.0, dropout=0.1, kl_loss_weight=0, exclude_neg=False, no_weights=False, lora=False, r=16, lora_alpha=32, lora_dropout=0.1, sampling_size=20, sampling_method='minority', cwes=['all'], langs=['all'], logging_steps=50, save_epochs=10, seed=2, data_dir='data_train_val', model_dir='../trained/', output_dir='../trained/starcoderbase-1b-safecoder', logger=<RootLogger root (INFO)>)
/home/xxx/conda-envs/xxx/lib/python3.10/site-packages/transformers/optimization.py:640: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
03/12/2025 14:27:50 - INFO - root -   ***** Running training *****
03/12/2025 14:27:50 - INFO - root -     Num samples = 31703
03/12/2025 14:27:50 - INFO - root -     Num epoch = 2
03/12/2025 14:27:50 - INFO - root -     Batch size= 1
03/12/2025 14:27:50 - INFO - root -     Total batch size (w. accumulation) = 16
03/12/2025 14:27:50 - INFO - root -     Gradient Accumulation steps = 16
03/12/2025 14:27:50 - INFO - root -     Total optimization steps = 3962
03/12/2025 14:27:50 - INFO - root -     Num val samples = 3443
03/12/2025 14:27:50 - INFO - root -     Num parameters = 1137207296
03/12/2025 14:27:50 - INFO - root -     Num trainable parameters = 117641216
03/12/2025 14:29:25 - INFO - root -   epochs: 1/2, steps: 50/3962, func: nan, pos: nan, neg: nan, 1%: 2h 4m 22s          
03/12/2025 14:30:58 - INFO - root -   epochs: 1/2, steps: 100/3962, func: nan, pos: nan, neg: nan, 2%: 2h 1m 22s    
```

I tried it multiple times and with my own datasets however this issue always occurs. Do you have any idea on this? Thank you very much!

**Comment by mveroe** [comment]
Thank you for your interest in our work! We did not encounter this issue, so I do not know from the top of my head what could be the cause. Can you please provide more detailed traces on when exactly the loss becomes nan? And if it is already the loss that evaluates to nan or the gradients are nan? Maybe try to look at all the relevant values in after each dataset sample.

**Comment by ljj430** [comment]
Thank you for your reply!

I inserted certain statements to print loss and other things in the function `step()` of `trainer.py`
The modified function is shown below:

```
   def step(self, batch):
        loss_dict = LossDict(self.loss_keys)

        sample_types, inputs, weights = batch
        inputs = inputs.to(self.model.device)
        shift_inputs = inputs[..., 1:]
        weights = weights.to(self.model.device)
        shift_weights = weights[..., 1:]
        
            
        
        outputs = self.model(inputs)
        

        shift_logits = outputs.logits[..., :-1, :]

        ### debug ###
        print(f"Checking inputs: NaN: {torch.isnan(inputs).any()}, Inf: {torch.isinf(inputs).any()}")
        print(f"Checking weights: NaN: {torch.isnan(weights).any()}, Inf: {torch.isinf(weights).any()}")
        print(f"Checking outputs logits: NaN: {torch.isnan(outputs.logits).any()}, Inf: {torch.isinf(outputs.logits).any()}")
        print(f"Checking shift_logits: NaN: {torch.isnan(shift_logits).any()}, Inf: {torch.isinf(shift_logits).any()}")
        print(f"shift_logits stats: min {shift_logits.min()}, max {shift_logits.max()}")
        print(f"shift_inputs stats: min {shift_inputs.min()}, max {shift_inputs.max()}")
        print(f"shift_weights stats: min {shift_weights.min()}, max {shift_weights.max()}")
        ### end ###
        
        loss_total = 0.0
        for sample_type in sample_types:
            if sample_type == FUNC:
                
                
                loss = token_weighted_loss('ce', shift_logits, shift_inputs, shift_weights)
                ### debug ###
                print("In FUNC of step()")
                print(f"loss: {loss}")
                print(f"loss.item(): {loss.item()}")
                
                print("*"*50)
                ### end ###
                loss_dict['func'].append(loss.item())
                loss_total += loss
            elif sample_type == GOOD:
                loss = self.args.loss_weight * token_weighted_loss('ce', shift_logits, shift_inputs, shift_weights)
                ### debug ###
                print("In GOOD of step()")
                print(f"loss: {loss}")
                print(f"loss.item(): {loss.item()}")
                print("*"*50)
                ### end ###
                loss_dict['pos'].append(loss.item())
                loss_total += loss
            elif sample_type == BAD:
                
                
                loss = self.args.loss_weight * token_weighted_loss('ul', shift_logits, shift_inputs, shift_weights)
                ### debug ###
                print("In BAD of step()")
                print(f"loss: {loss}")
                print(f"loss.item(): {loss.item()}")
                print("*"*50)
                ### end ###
                loss_dict['neg'].append(loss.item())
                loss_total += loss
            else:
                assert False

            ### debug ###
            if sample_type == GOOD or sample_type == BAD:
                print("Checking kl_loss_weight in step()")
                print(f"args.kl_loss_weight: {self.args.kl_loss_weight}")
                print(f"loss.item(): {loss.item()}")
                print("*"*50)
            ### end ###
            if (sample_type == GOOD or sample_type == BAD) and self.args.kl_loss_weight > 0:
                with torch.no_grad():
                    ref_outputs = self.ref_model(inputs)
                shift_ref_log_probs = F.log_softmax(ref_outputs.logits[..., :-1, :], dim=-1)
                shift_log_probs = F.log_softmax(shift_logits, dim=-1)
                loss = self.args.kl_loss_weight * token_weighted_loss('kl', shift_log_probs, shift_ref_log_probs, 1-shift_weights) / 1000
                
                ### debug ###
                print("In GOOD or BAD with kl_loss of step()")
                print(f"loss: {loss}")
                print(f"loss.item(): {loss.item()}")
                print("*"*50)
                ### end ###
                loss_dict['kl'].append(loss.item())
                loss_total += loss

        return loss_total, loss_dict
```
Then I ran `python train.py --pretrain_name starcoderbase-1b --output_name starcoderbase-1b-safecoder-1 --datasets evol sec-desc sec-new-desc --logging_steps=10`

Here is the output, the beginning is normal but after a few steps it becomes NaN (the last two outputs):
```
(/home/xxx/conda-envs/xxx) xxx@xxx:~/xxx/SafeCoder/scripts$ python train.py --pretrain_name starcoderbase-1b --output_name starcoderbase-1b-safecoder-1 --datasets evol sec-desc sec-new-desc --logging_steps=10
The `load_in_4bit` and `load_in_8bit` arguments are deprecated and will be removed in the future versions. Please, pass a `BitsAndBytesConfig` object in `quantization_config` argument instead.
03/19/2025 16:12:24 - INFO - accelerate.utils.modeling -   We will use 90% of the memory on device 0 for storing the model, and 10% for the buffer to avoid OOM. You can set `max_memory` in to a higher value to use more memory (at your own risk).
03/19/2025 16:13:05 - INFO - root -   number of sec samples before upsampling: 1848
03/19/2025 16:13:05 - INFO - root -   number of sec samples after upsampling: 2468
03/19/2025 16:13:06 - INFO - root -   Training args Namespace(output_name='starcoderbase-1b-safecoder-1', datasets=['evol', 'sec-desc', 'sec-new-desc'], pretrain_name='starcoderbase-1b', loss_weight=1.0, sven=False, num_train_epochs=2, learning_rate=2e-05, max_num_tokens=1024, batch_size=1, grad_acc_steps=16, weight_decay=0.01, adam_epsilon=1e-08, warmup_steps=0, max_grad_norm=1.0, dropout=0.1, kl_loss_weight=0, exclude_neg=False, no_weights=False, lora=False, r=16, lora_alpha=32, lora_dropout=0.1, sampling_size=20, sampling_method='minority', cwes=['all'], langs=['all'], logging_steps=10, save_epochs=10, seed=2, data_dir='../data_train_val', model_dir='../trained/', output_dir='../trained/starcoderbase-1b-safecoder-1', logger=<RootLogger root (INFO)>)
/home/xxx/conda-envs/xxx/lib/python3.10/site-packages/transformers/optimization.py:640: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
03/19/2025 16:13:06 - INFO - root -   ***** Running training *****
03/19/2025 16:13:06 - INFO - root -     Num samples = 31703
03/19/2025 16:13:06 - INFO - root -     Num epoch = 2
03/19/2025 16:13:06 - INFO - root -     Batch size= 1
03/19/2025 16:13:06 - INFO - root -     Total batch size (w. accumulation) = 16
03/19/2025 16:13:06 - INFO - root -     Gradient Accumulation steps = 16
03/19/2025 16:13:06 - INFO - root -     Total optimization steps = 3962
03/19/2025 16:13:06 - INFO - root -     Num val samples = 526
03/19/2025 16:13:06 - INFO - root -     Num parameters = 1137207296
03/19/2025 16:13:06 - INFO - root -     Num trainable parameters = 117641216
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -20.40625, max 29.703125
shift_inputs stats: min 0, max 46241
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 0.994140625
loss.item(): 0.994140625
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -18.53125, max 23.65625
shift_inputs stats: min 0, max 44314
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.626953125
loss.item(): 1.626953125
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -19.25, max 26.21875
shift_inputs stats: min 0, max 49125
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 2.38671875
loss.item(): 2.38671875
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -17.421875, max 23.109375
shift_inputs stats: min 20, max 48031
shift_weights stats: min 0, max 1
In GOOD of step()
loss: 1.978515625
loss.item(): 1.978515625
**************************************************
Checking kl_loss_weight in step()
args.kl_loss_weight: 0
loss.item(): 1.978515625
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -18.375, max 29.671875
shift_inputs stats: min 0, max 48588
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.265625
loss.item(): 1.265625
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -20.15625, max 30.25
shift_inputs stats: min 0, max 49100
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.3525390625
loss.item(): 1.3525390625
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -18.296875, max 30.328125
shift_inputs stats: min 0, max 44527
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.357421875
loss.item(): 1.357421875
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -19.125, max 27.765625
shift_inputs stats: min 0, max 39290
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.4560546875
loss.item(): 1.4560546875
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -19.1875, max 24.671875
shift_inputs stats: min 0, max 48369
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 2.087890625
loss.item(): 2.087890625
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -17.9375, max 25.578125
shift_inputs stats: min 0, max 36808
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 0.99267578125
loss.item(): 0.99267578125
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -19.203125, max 24.625
shift_inputs stats: min 0, max 46690
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.173828125
loss.item(): 1.173828125
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -18.40625, max 22.84375
shift_inputs stats: min 0, max 38302
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 0.81494140625
loss.item(): 0.81494140625
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -19.75, max 31.8125
shift_inputs stats: min 0, max 48299
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.0703125
loss.item(): 1.0703125
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -17.6875, max 23.21875
shift_inputs stats: min 0, max 36808
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.1259765625
loss.item(): 1.1259765625
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -18.75, max 24.78125
shift_inputs stats: min 0, max 48919
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.787109375
loss.item(): 1.787109375
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: False, Inf: False
Checking shift_logits: NaN: False, Inf: False
shift_logits stats: min -18.75, max 26.703125
shift_inputs stats: min 0, max 48852
shift_weights stats: min 0, max 1
In FUNC of step()
loss: 1.306640625
loss.item(): 1.306640625
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: True, Inf: False
Checking shift_logits: NaN: True, Inf: False
shift_logits stats: min nan, max nan
shift_inputs stats: min 0, max 46245
shift_weights stats: min 0, max 1
In FUNC of step()
loss: nan
loss.item(): nan
**************************************************
Checking inputs: NaN: False, Inf: False
Checking weights: NaN: False, Inf: False
Checking outputs logits: NaN: True, Inf: False
Checking shift_logits: NaN: True, Inf: False
shift_logits stats: min nan, max nan
shift_inputs stats: min 23, max 41964
shift_weights stats: min 0, max 1
In BAD of step()
loss: nan
loss.item(): nan
**************************************************
```

Do you have any idea about this? Thank you very much for your help and please let me know if you need further information.

**Comment by mveroe** [comment]
Hey, thanks for the traces. This looks quite strange to me, as it seems that the first time NaN is found is coming directly from the outputs of the model. Can you maybe try to check the gradient right before the first time NaN occurs and also the values right after projection? Normally the classification head of the model should not produce NaNs.

**Comment by ljj430** [comment]
Hi, thank you for your help! I think I’ve resolved the NaN issue. It was caused by using mixed precision—specifically, loading the model with `torch.float16` instead of `torch.float32` to save GPU memory. After switching to `torch.float32`, the loss behaves normally.
